xdem/terrain.py

```diff
@@ -0,0 +1,232 @@
+    # Validate and format the inputs
+    if isinstance(attribute, str):
+        attribute = [attribute]
+
+    choices = ["slope", "aspect", "hillshade"]
+    for attr in attribute:
+        if attr not in choices:
+            raise ValueError(f"Attribute '{attr}' is not supported. Choices: {choices}")
+
+    if not isinstance(resolution, Sized):
+        resolution = (float(resolution), float(resolution))
+
+    if (hillshade_azimuth < 0.0) or (hillshade_azimuth > 360.0):
+        raise ValueError(f"Azimuth must be a value between 0 and 360 degrees (given value: {hillshade_azimuth})")
+    if (hillshade_altitude < 0.0) or (hillshade_altitude > 90):
+        raise ValueError("Altitude must be a value between 0 and 90 degress (given value: {altitude})")
+    if (hillshade_z_factor < 0.0) or not np.isfinite(hillshade_z_factor):
+        raise ValueError(f"z_factor must be a non-negative finite value (given value: {hillshade_z_factor})")
+
+    dem_arr = xdem.spatial_tools.get_array_and_mask(dem)[0]
+
+    # Calculate the gradient, from which all products are derived.
+    x_gradient, y_gradient = np.gradient(dem_arr)
+
+    # Normalize by the radius of the resolution to make it resolution variant.
+    x_gradient /= resolution[0]
+    y_gradient /= resolution[1]
+
+    # Initialize the terrain_attributes as tiny empty ndarrays
+    # This is to make the syntax below easier. They won't be used if they are not filled with something else.
+    terrain_attributes = {"hillshade": np.array([np.nan]), "aspect": np.array([np.nan]), "slope": np.array([np.nan])}
+
+    # Check which products should be made
+    make_aspect = any(attr in attribute for attr in ["aspect", "hillshade"])
+    make_slope = any(attr in attribute for attr in ["slope", "hillshade"])
+    make_hillshade = "hillshade" in attribute
+
+    if make_aspect:
+        terrain_attributes["aspect"] = np.arctan2(-x_gradient, y_gradient)
+
+    if make_slope:
+        terrain_attributes["slope"] = np.pi / 2.0 - np.arctan(np.sqrt(x_gradient ** 2 + y_gradient ** 2))
+
+    if make_hillshade:
+        # If a different z-factor was given, recaculate the slopemap with exaggerated gradients.
+        if hillshade_z_factor != 1.0:
+            slopemap = np.pi / 2.0 - np.arctan(np.sqrt((x_gradient * hillshade_z_factor) ** 2 + (y_gradient * hillshade_z_factor) ** 2))
+        else:
+            slopemap = terrain_attributes["slope"]
+
+        azimuth_rad = np.deg2rad(360 - hillshade_azimuth)
+        altitude_rad = np.deg2rad(hillshade_altitude)
+        terrain_attributes["hillshade"] = np.clip(
+            255
+            * (
+                np.sin(altitude_rad) * np.sin(slopemap)
+                + np.cos(altitude_rad)
+                * np.cos(terrain_attributes["slope"])
+                * np.cos((azimuth_rad - np.pi / 2.0) - terrain_attributes["aspect"])
+            ),
+            0,
+            255,
+        ).astype("float32")
+
+    # Convert the unit if wanted.
+    if degrees:
```

Comment:
It is in radians, since aspect and slope are using `np.arctan2` and `np.arctan`, respectively. I haven't checked from where it begins, however (like if 0 is N, pi is south, and so on)!

Reply:
Yes this is what I meant for "where it begins"! 
I think you can use the same calculation you did for degrees, but directly in radians. And then if `degrees=degrees` convert directly into degrees without any other calculation.

Reply:
Fixed in 9b9e3c3